projects/src/transmitter.cc

```diff
@@ -0,0 +1,157 @@
+#include "packet/serialize.h"
+#include "usbconfig.h"
+
+// This define can be useful when sizing message mailbox and memory pools
+#define ASSERT_MESSAGE_MEMORY_LIMIT FALSE
+
+namespace message {
+Transmitter::Transmitter() :
+m_thread(nullptr),
+m_bytes_written(0) {
+    initialize();
+
+    // Initialize a serial-over-USB CDC driver.
+    sduObjectInit(&SDU1);
+    sduStart(&SDU1, &serusbcfg);
+
+    // Activate the USB driver and then the USB bus pull-up on D+.
+    // Note, a delay is inserted in order to not have to disconnect the cable
+    // after a reset.
+    board_usb_lld_disconnect_bus();   //usbDisconnectBus(serusbcfg.usbp);
+    chThdSleepMilliseconds(1500);
+    usbStart(serusbcfg.usbp, &usbcfg);
+    board_usb_lld_connect_bus();      //usbConnectBus(serusbcfg.usbp);
+
+    // Block USB device until ready
+    while ((SDU1.config->usbp->state != USB_ACTIVE) || (SDU1.state != SDU_READY)) {
+        chThdSleepMilliseconds(10);
+    }
+}
+
+void Transmitter::start(tprio_t priority) {
+    m_thread = chThdCreateStatic(m_wa_transmitter_thread,
+            sizeof(m_wa_transmitter_thread), priority,
+            transmitter_thread_function, this);
+}
+
+BicyclePoseMessage* Transmitter::alloc_pose_message() {
+    void* msg = chPoolAlloc(&m_pose_message_pool);
+#if ASSERT_MESSAGE_MEMORY_LIMIT
+    chDbgAssert(msg != nullptr, "Increase transmitter POSE_MESSAGE_POOL_SIZE");
+#endif
+    return static_cast<BicyclePoseMessage*>(msg);
+}
+
+void Transmitter::free_pose_message(BicyclePoseMessage* msg) {
+    chPoolFree(&m_pose_message_pool, reinterpret_cast<void*>(msg));
+}
+
+msg_t Transmitter::transmit_async(BicyclePoseMessage* msg) {
+    msg_t status = chMBPost(&m_message_mailbox, reinterpret_cast<msg_t>(msg), TIME_IMMEDIATE);
+#if ASSERT_MESSAGE_MEMORY_LIMIT
+    chDbgAssert(status == MSG_OK, "Increase transmitter MAILBOX_SIZE");
+#endif
+    return status;
+}
+
+SimulationMessage* Transmitter::alloc_simulation_message() {
+    void* msg = chPoolAlloc(&m_simulation_message_pool);
+#if ASSERT_MESSAGE_MEMORY_LIMIT
+    chDbgAssert(msg != nullptr, "Increase transmitter SIMULATION_MESSAGE_POOL_SIZE");
+#endif
+    return static_cast<SimulationMessage*>(msg);
+}
+
+void Transmitter::free_simulation_message(SimulationMessage* msg) {
+    chPoolFree(&m_simulation_message_pool, reinterpret_cast<void*>(msg));
+}
+
+msg_t Transmitter::transmit_async(SimulationMessage* msg) {
+    msg_t status = chMBPost(&m_message_mailbox, reinterpret_cast<msg_t>(msg), TIME_IMMEDIATE);
+#if ASSERT_MESSAGE_MEMORY_LIMIT
+    chDbgAssert(status == MSG_OK, "Increase transmitter MAILBOX_SIZE");
+#endif
+    return status;
+}
+
+void Transmitter::transmit(SimulationMessage* msg) {
+    if (m_thread != nullptr) {
+        chSysHalt("Synchronization is not implemented and this function is not thread safe");
+    }
+
+    encode_message(reinterpret_cast<msg_t>(msg));
+    transmit_packet();
+    free_simulation_message(msg);
+}
+
+void Transmitter::initialize() {
```

Comment:
If this function is only called from the constructor, just put the code in the constructor.